service/storage_service.cc

```diff
@@ -4845,6 +4845,24 @@ future<raft_topology_cmd_result> storage_service::raft_topology_cmd_handler(shar
                     // we already did read barrier above
                     result.status = raft_topology_cmd_result::command_status::success;
                 break;
+                case raft_topology_cmd::command::barrier_and_drain:
+                    co_await container().invoke_on_all([param = cmd.param] (storage_service& ss) -> future<> {
+                        const auto requested_version = holds_alternative<topology::version_t>(param)
```

Comment:
IMO there is no need to pass the version we want to drain explicitly. If we are here it means current metadata version is up-to-date and we need to wait for all earlier versions to complete.

Reply:
You first store the same state in to different places (rpc call and topology) and then you try to make sure they do not go out of sync while you have no a way to recover if they do. What is the point? The thing is all other rpc calls here already use the way I described. I think we should continue at least for consistency.

Reply:
OTOH we need to protect from commands been re-ordered. We already have a protection from executing a command from outdated coordinator by having the term in the cmd and checking it before executing a command. What we need is to add a counter that is monotonically increased for each term and remember the id of the last command. The cmd can be executed only if it's id is higher than the previous one otherwise it is dropped. This way we can fully rely on a leader state and a follower state to be in sync and to not add different checks in each and every cmd.

Reply:
ok